core/templates/services/playthrough.service.ts

```diff
@@ -107,26 +109,26 @@ class CyclicStateTransitionsTracker {
    * Finally, the path of visited states is reset to a value of [ N ], in hopes
    * that the exact same cycle is built enough times to be considered an issue.
    *
-   * Note that 1-cycles (N -> N) are ignored. When a 1-cycle is discovered, the
-   * tracker still resets itself to stay consistent with the behavior taken for
-   * other cycles.
+   * Note that 1-cycles (N -> N) are exceptional and ignored completely. They do
+   * not reset the tracker and do not increment the tracker's cycle occurrences.
    */
   recordStateTransition(destStateName: string): void {
-    if (!this.pathOfVisitedStates.includes(destStateName)) {
-      this.pathOfVisitedStates.push(destStateName);
+    if (this.currStateName() === destStateName) {
       return;
     }
-    if (this.currStateName() !== destStateName) {
+    if (!this.pathOfVisitedStates.includes(destStateName)) {
+      this.pathOfVisitedStates.push(destStateName);
+    } else {
       const cycleOfVisitedStates = (
         this.makeCycle(this.pathOfVisitedStates.indexOf(destStateName)));
       if (angular.equals(this.cycleOfVisitedStates, cycleOfVisitedStates)) {
-        this.numLoops += 1;
+        this.cycleOccurrences += 1;
```

Comment:
I was thinking more about the logic here. Would it never detect that the cycle A --> B --> C --> A is happening twice in A --> B --> C --> A --> B --> C --> B --> C --> A --> B --> C --> A? Because:

- after second visit to A, the cycle is A --> B --> C --> A, it's recorded as 1 and visitedStates is compressed to [A]
- after third visit to B, cycle is B --> C --> B, it's recorded as 1 and visitedStates is compressed to [B]
- after fourth visit to B, cycle is B --> C --> A --> B, it's recorded as 1 and visitedStates is compressed to [B]
- after last visit to A, cycle is A --> B --> C --> A, and the earlier record is not remembered so it's recorded as cycleOccurrences = 1

That's a problem, right? (Or have I misread this?)

Reply:
Right, that's the intended behavior. Only consecutive cycles are considered issues, disjoint cycles are not. The context to keep in mind is that this issue is only recorded when it leads directly to a quit, so having looped a few states earlier in the exploration does not fit in with that context.

Reply:
But in this case the learner is clearly having trouble with states A, B, C. I can understand your point in the event that they go to D, E, F and then later come back to this cluster, but the playthrough provided above suggests that there is a problem in A, B, C, right?

So maybe we should be doing something which takes into account the *set* of states in the cycle...?

Reply:
Just to be clear, it sounds like there are two separate functionality requests here:
1. Maintain full history of discovered cycles, return an issue when *any* of those cycles have occurred enough times.
2. Extend "equality" of cycles to include rotations (i.e., "BCAB" === "ABCA").

Is that right?

Reply:
Have gone ahead and assumed that to be the case -- Done.

Reply:
Hmm. I'm not sure what the exact requirements are, but the general idea I was thinking of was that

A --> B --> C --> A --> B --> C --> B --> C --> A --> B --> C --> A

should be counted as 3 occurrences of an A --> B --> C --> A cycle (do you think that is sensible?). So, perhaps not quite (2).

Also, with (2), aren't you in danger of counting overlapping cycles? For example, A --> B --> C --> A --> B has only one cycle really, but if you have ABCA and BCAB being isomorphic, wouldn't it look as though there are two?